Add back previous run and profile updates

thorrester opened the pull request:
## Pull Request

### Short Summary
- Adds back `previous_run` from the `drift_profile` table and use it in the limit timestamp clause.
- Noticed the drift array needed to be transposed to generate alerts so that the shape of the drift array matches that of the generated drift value array
- Making sure to feed the features to `generate_alerts` in the same order in which they are read from the sql table.
- Made all drift profiles `not active` by default in order to avoid a proliferation of jobs.
- Users will opt-in to turning on monitoring when they set up their apis via the new `profile/status` route 
    - We can add a cli command to `opsml-cli` + sdk integration into `opsml` later.
 - Added @russellkemmit script for populating the db
 - Added tests for alert and dispatch
 - Removed the nested loop for the background alert job. Keep loop at the parent-level, but create more workers per server

thorrester commented on src/alerts/drift.rs:
```diff
@@ -43,36 +43,51 @@ impl DriftExecutor {
         let drift = Monitor::new().calculate_drift_from_sample(
             &feature_keys,
-            &nd_feature_arr.view(),
+            &nd_feature_arr.t().view(), // need to transpose because calculation is done at the row level across each feature
```
This is to ensure feature len matches what is expected since we loop over each row and then loop over each feature in a row

thorrester commented on src/alerts/drift.rs:
```diff
@@ -88,61 +103,64 @@ impl DriftExecutor {
-            let mut transaction: sqlx::Transaction<Postgres> = self.db_client.pool.begin().await?;
-            // Get drift profile
-            let profile = PostgresClient::get_drift_profile(&mut transaction)
+        let mut transaction: sqlx::Transaction<Postgres> = self.db_client.pool.begin().await?;
```
removed the loop here since it's already running in a loop

russellkemmit replied:
This loop is needed to ensure all profiles are processed.

thorrester commented on src/api/handler.rs:
```diff
@@ -124,3 +124,38 @@ pub async fn insert_drift_profile(
     }
 }
+
+pub async fn update_drift_profile_status(
```
new logic for updating a profiles status. Once it's activated to `true`, the background tasks should pick it up

thorrester commented on src/main.rs:
```diff
@@ -17,7 +17,8 @@ use std::sync::Arc;
 use tracing::{error, info};
 
-const NUM_WORKERS: usize = 5;
+const NUM_WORKERS: usize = 3;
```
I think we can make these in env vars later on

thorrester commented on src/main.rs:
```diff
@@ -88,18 +89,21 @@ async fn start_main_server() -> Result<(), anyhow::Error> {
-            interval.tick().await;
-            if let Err(e) = drift_executor.execute().await {
-                error!("Drift Executor Error: {e}")
+    for i in 0..NUM_DISPATCHERS {
```
changed to a loop to create n background scheduler tasks

russellkemmit replied:
Need this loop to ensure we process all active profiles

russellkemmit replied:
Hell Yes!!!

thorrester commented on src/sql/postgres.rs:
```diff
@@ -154,12 +155,12 @@ impl PostgresClient {
             repository: drift_profile.config.repository.clone(),
             version: drift_profile.config.version.clone(),
             profile: serde_json::to_string(&drift_profile).unwrap(),
+            active: false,
```
user must opt-in to monitoring or else anytime someone create a drift profile, the monitoring job will pick it up

thorrester commented on tests/drift_integration.rs:
```diff
@@ -0,0 +1,119 @@
+use sqlx::{Postgres, Row};
+mod test_utils;
+
+#[tokio::test]
```
test to use new populate script

russellkemmit replied:
Let's GO!!

thorrester commented:
Added loop back + made some additional changes
- Better naming for functions
- Additional error handling if drift array is empty
- Added sleep for when there's nothing left to process in order to add a little bit of a delay
- Changed the naming in the makefile
